### PR TITLE
Skip CXformPushGbBelowSetOp when arity > threshold

### PR DESCRIPTION
**Pull request by hardikar**
In case of queries containing a UNION over a large number of children,
CXformPushGbBelowSetOp pushes the Group By op to each child, creating
that many new groups in the memo. While not a problem by itself, it
exponentially increases the number of bindings for other xforms that use
CPatternTree to extract it. When extracting the bindings, each child of
UNION can now be selected with and without the group by. The higher the
number of such children, the more the number of possibilities, the
higher the optimization time.

An example of xform that is affected: CXformLeftSemiApply2LeftSemiJoin.

To prevent this exponential increase in binding extraction, restrict
when the CXformPushGbBelowSetOp may be executed by a configurable
parameter.

The intention here is also to create a GUC in GPDB to be able to control this
parameter. Also we need to change all the MDPs to use the new default value.

5X PR: https://github.com/greenplum-db/gpdb/pull/7897
master PR: https://github.com/greenplum-db/gpdb/pull/7898

### CODE REVIEW

**Comment by bhuvnesh2703** [comment]
I am fine with this approach, as i don't see what else we can do when we have a high number of union childs..Potentially, the only caveat is in some cases the execution time will be really better if we push the group by to each of the child instead of doing it on top of the union all..
What value do you think is acceptable based on optimization time?

**Comment by vraghavan78** [comment]
> I am fine with this approach, as i don't see what else we can do when we have a high number of union childs..Potentially, the only caveat is in some cases the execution time will be really better if we push the group by to each of the child instead of doing it on top of the union all..
> What value do you think is acceptable based on optimization time?

It is hard to predict which place it will benefit before the fact since this is a logical optimization.

**Comment by hardikar** [comment]
Running a query with varying UNIONS inside a subquery, the time to run the EXPLAIN query with 5 UNIONs is about the same as time to run any query with CXformPushGbBelowSetOp disable. So 6 could be reasonable number for the threshold, if we want to give an illusion of continuity of planning time. However, I feel like from a SQL writer's perspective, it is too low. 10 is looking increasingly like a better compromise. But any number is going to be only a guess (or even overfitted to a specific query). Ultimately, this should be configurable.

**Comment by chrishajas** [comment]
This seems fine to me. With 10+ unions, the optimization time will start taking minutes+, which doesn't seem acceptable in many circumstances. I'd probably err on the side of setting it higher (10) rather than 5, just so we don't accidentally make execution times worse. I assume there's no plan changes on the workload pipeline (once it's working again)?

File: data/dxl/minidump/ManyTextUnionsInSubquery.mdp [hunk 1]
```diff
@@ -0,0 +1,1059 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+  <dxl:Comment><![CDATA[
```
Comment (zellerh) [inline data/dxl/minidump/ManyTextUnionsInSubquery.mdp]: Might be good to put a comment here that you expect the plan not to have the groupby pushed down below the union.

**Comment by hardikar** [comment]
Addressed Hans' comment. And @chrishajas workload schedule 1 looks good. Pushing this in now. Thanks for the review!

**Pull request merged**